### ISSUE TITLE
Removal of newlines breaks some markup

### ISSUE THREAD

**Issue by goofballLogic** [markcellus/html-express-js#57]
For some markup (e.g. `pre` blocks) the line breaks matter. Currently the library strips out `\n` and `\r`. Can this somehow be configurable / optional?

**Comment by markcellus** [comment]
Hey, thanks for filing the issue. Yeah, the library strips out these characters since they slow down parsing and browsers generally discard them anyway. But before changing any behavior, can you describe your specific use-case that requires these characters in HTML?

**Comment by goofballLogic** [comment]
so, for example, I have an app where I render normal HTML for most of the page, but I also have a block which renders content based on newlines and pre-determined white-space. (In this case it's actually a representation of data in JSON format).

The use case is that an administrative user can see the configuration of plug-ins as specified in JSON format. The site rendering this configuration doesn't know what it means, so it just displays it "raw":

```
<pre>
   {
      "rule": "dt-spend-down",
      "test": {
         "type": "max-age",
         "age": "15"
      }
   }
</pre>```

Currently, this gets fubarred by the removal of line breaks.

**Comment by goofballLogic** [comment]
Also  - a note - I agree that removal of line breaks is generally a good idea. I think what's needed is a way to "opt-out" of the line-break removal. Your current mechanism which strips whitespace from template parameters immediately doesn't allow this.

**Comment by markcellus** [comment]
Hmm is there any reason why the data in `pre` can just be passed to the client and in your client you parse that into JSON-formatted object via `JSON.parse()`? It would also be more secure by parsing/formatting the JSON on the client instead of trying to parse it through the HTML response of the server.

**Comment by goofballLogic** [comment]
Plus, I think any decent view engine should be compatible with rendering the full set of HTML tags as a basic minumum. A `<pre>` tag (or any other tag rendered with CSS indicating white-space: pre) requires control over the whitespace.

**Comment by markcellus** [comment]
> I think any decent view engine should be compatible with rendering the full set of HTML tags as a basic minumum

This view engine does just that. It's meant to render HTML--not formatted JSON unfortunately, which seems to be required for your use-case. 

But maybe I'm missing something. Whether it's server-side rendering or client-side rendering shouldn't matter in this context. Why exactly can't the JSON be formatted after the first (and main) render?

**Comment by markcellus** [comment]
Also, just wondering as I'm thinking about this. If the JSON is already available, what's the reason for rendering it as HTML just to get rid of the `<pre>` tags? Can the server just respond with the JSON string without using the `html()` in this view engine?

**Comment by goofballLogic** [comment]
Whitespace this is HTML. Lets ignore the JSON for a second. Lets just go back to raw text. I'd like to render this HTML output:
```
<pre>
   This is the first line, indented.
   And, this is the second line.
</pre>
```

**Comment by goofballLogic** [comment]
Here's an example: https://jsfiddle.net/r6nse2m9/

**Comment by goofballLogic** [comment]
And here is Pug (aka Jade) documentation of how.they do it: https://pugjs.org/language/plain-text.html#whitespace-control

**Comment by markcellus** [comment]
Thanks for that information, I think I understand now. The `<pre>` tag should definitely work as it was intended. 

Instead of making this an option in the `html()` function, is there any reason this shouldn't just be a global setting for all usage of the package? I'd assume users would expect they'd want the same behavior whenever `html()` is used.

**Comment by goofballLogic** [comment]
well i'd assume that people want whitespace stripped everywhere except for inside the <pre> tag wouldn't they? (or like Pug does it - whitespace between elements stripped, but not inside them?

**Comment by markcellus** [comment]
As your example shows, almost any tag can be made to have `pre` style formatting via CSS though.

**Comment by goofballLogic** [comment]
Updated jsfiddle to add those two use cases: https://jsfiddle.net/9mb87xw2/

**Comment by markcellus** [comment]
Great, thanks for that. I'll put up a PR to make whitespace removal opt out at a global level. Any whitespace removal needed at the response level can easily be done manually. 👍

**Comment by goofballLogic** [comment]
Can you do it using a strategy pattern? That way I could try a PR for a pug-style version (or even a case-by-case opt-out version) ?

**Comment by markcellus** [comment]
Hey I just ditched all whitespace removal in the package. I realized I can remove any whitespace I need explicitly in my app to keep it performant. If anyone needs whitespace removal, it's pretty trivial to do, so didn't really see much of a need to include this as an option anymore. But if you think there needs to still be an option for it, feel free to open up another issue and we can discuss more. Thanks!
